core/drivers/dcp/imx_dcp_utils.c

```diff
@@ -0,0 +1,185 @@
+#define OF_MEM_HDR(p) (struct mem_hdr *)((uint8_t *)(p) - (MEM_HDR_SIZE))
+
+/*
+ * Read the system cache line size.
+ * Get the value from the ARM system configuration register
+ */
+static uint32_t read_cacheline_size(void)
+{
+	uint32_t value = 0;
+
+#ifdef ARM64
+	value = read_ctr_el0();
+#else
+	value = read_ctr();
+#endif /* ARM64 */
+	value = CTR_WORD_SIZE
+		<< ((value >> CTR_DMINLINE_SHIFT) & CTR_DMINLINE_MASK);
+	DMSG("System Cache Line size = %" PRIu32 " bytes", value);
+
+	return value;
+}
+
+/*
+ * Allocate an area of given size in bytes and set it to zeor. Add the memory
+ * allocator information in the newly allocated area.
+ * Return a cacheline aligned buffer.
+ *
+ * @size   Size in bytes to allocate
+ */
+static void *dcp_alloc_memalign(size_t size)
+{
+	struct mem_hdr *hdr = NULL;
+	vaddr_t ret_addr = 0;
+	void *ptr = NULL;
+	size_t alloc_size = 0;
+	uint32_t cacheline_size = 0;
+
+	/*
+	 * The mem_hdr header is added just before the returned
+	 * buffer address
+	 *
+	 * --------------
+	 * |  mem_hdr   |
+	 * --------------
+	 * |  Buffer    |
+	 * --------------
+	 */
+	if (ADD_OVERFLOW(size, MEM_HDR_SIZE, &alloc_size))
+		return NULL;
+
+	/*
+	 * Buffer must be aligned on a cache line:
+	 *  - Buffer start address aligned on a cache line
+	 *  - End of Buffer inside a cache line.
+	 *
+	 * If area's (mem info + buffer) to be allocated size is
+	 * already cache line aligned add a cache line.
+	 *
+	 * Because Buffer address returned is moved up to a cache
+	 * line start offset, add a cache line to full area allocated
+	 * to ensure that end of the working buffer is in a cache line.
+	 */
+	cacheline_size = read_cacheline_size();
+	if (size == cacheline_size) {
+		if (ADD_OVERFLOW(alloc_size, cacheline_size, &alloc_size))
+			return NULL;
+	}
+
+	if (ADD_OVERFLOW(cacheline_size, ROUNDUP(alloc_size, cacheline_size),
+			 &alloc_size))
+		return NULL;
+
+	ptr = calloc(1, alloc_size);
+
+	if (!ptr) {
+		EMSG("alloc Error - NULL");
+		return NULL;
+	}
+
+	/* Calculate the return buffer address */
+	ret_addr = (vaddr_t)ptr + MEM_HDR_SIZE;
+	ret_addr = ROUNDUP(ret_addr, cacheline_size);
+
+	/* Add the mem_hdr header */
+	hdr = OF_MEM_HDR(ret_addr);
+	hdr->addr = ptr;
+	hdr->size = alloc_size;
+
+	return (void *)ret_addr;
+}
+
+/*
+ * Free aligned allocated area
+ *
+ * @ptr   area to free
+ */
+static void dcp_free_memalign(void *ptr)
+{
+	struct mem_hdr *hdr = NULL;
+
+	if (!ptr)
+		return;
+
+	hdr = OF_MEM_HDR(ptr);
+
+	free(hdr->addr);
+}
+
+TEE_Result dcp_calloc_align_buf(struct align_buf *buf, size_t size)
+{
+	if (!buf) {
+		EMSG("Error, buf is null");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	buf->data = dcp_alloc_memalign(size);
+	if (!buf->data)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	buf->paddr = virt_to_phys(buf->data);
+
+	if (!buf->paddr) {
+		dcp_free(buf->data);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	buf->size = size;
+
+	return TEE_SUCCESS;
+}
+
+void dcp_free(void *ptr)
```

Comment:
It would be better symmetry if this function operated on a `struct align_buf` too.

Reply:
I agree, I will fix that